### DISCUSSION TITLE
General questions regarding zcompiling

### DISCUSSION THREAD

**Discussion by rieje** [zimfw/zimfw#547] [Q&A] [answered]
Disclaimer: not really zimfw-related, more zsh in general and I'm pretty new to zsh:

Should autoload functions and completion be zcompiled? I took a glance at zcompile implementations between some frameworks and plugin managers and have a few questions:

* I haven't come across any that includes autoload functions (unless it has the suffix `.zsh` but glancing over at random dotfiles, the naming scheme seems to be just the name of the function without the suffix). It also doesn't include completion files. Should these be zcompiled or handled? I understand it's probably not within the scope of plugin managers. I do see zgenom compiling for files with [zsh shebang](https://github.com/jandamm/zgenom/blob/main/functions/zgenom-compile#L39) which are included in their function files but autoload function files don't typically include this (I guess it's a good idea to?). 

* Would it make sense to compile multiple files into one .zwc file to reduce the number of files every shell init needs to read (or one per "module"? zcompile is usually instant on personal config so I'm not sure if there are introduced downsides. Checking if any one of the files in a .zwc digest was updated to update the digest might not be as straightforward though. I guess it's trading slightly longer init for technically better runtime (I guess doing it at .zlogin once per login session make most sense. I see some frameworks do something similar like`compinit` vs. `compinit -C` once per login) This idea is not new, can't remember where I came across it though.

* Zcompiling zsh builtin functions and scripts at /usr/share/zsh/{functions,scripts,site-functions} would be beneficial, right? I've never seen these covered in any tips for optimization (I don't except frameworks or plugin managers to interact with system files but that's where zsh's builtin functions and completions are defined).

Much appreciated.

**Comment by ericbn** [comment, marked as answer]
@rieje, autoloaded functions can be zcompiled into digest files. The completion functions are also autoloaded (by compinit), so they can potentially also be zcompiled into digest files. We used to create these digest files, but actually no performance gain was observed. We removed this at commit ae3e71cb84a538bad400ecca881be4ef5add5f33.

Also, function digest files would not impact the startup time significantly, as autoloaded functions are by definition lazy loaded.

**Reply by enoryw** [reply]
Hijacking this recent thread for clarification:

* Digest files are specifically .zwc files with *multiple* functions and it would be not be problematic (and marginally beneficial) to compile for completion/autoloaded files individually, i.e. one .zwc per file?
 
* Why is [zrecompile preferred over zcompile](https://github.com/zimfw/zimfw/issues/86) (`zcompare`)? The latter is faster, doesn't require loading an additional module, and doesn't litter filesystem with *.old files. Is the fact the zrecompile is native the only reason?

Thanks!
